### PR TITLE
[ASA-777][RV] relax validation rules for rejection text box

### PR DESCRIPTION
**Pull request by rvijayan**

### CODE REVIEW

File: app/uk/gov/hmrc/agentoverseasapplicationmaintainerfrontend/validators/CommonValidators.scala [hunk 1]
```diff
@@ -24,10 +24,10 @@ import uk.gov.hmrc.agentoverseasapplicationmaintainerfrontend.models.RejectReaso
 object CommonValidators {
 
   private val ApplicationReferenceRegex = """^[a-zA-Z0-9 ]*$"""
-  private val ReasonRegex = "^[A-Za-z0-9 \\-,.&']*$"
+  private val ReasonRegex = "^[A-Za-z0-9 \\-&.,_?:;!'@()\\[\\]\\\"]*$"
```
Comment (johnrevill) [inline app/uk/gov/hmrc/agentoverseasapplicationmaintainerfrontend/validators/CommonValidators.scala]: underscores are allowed in this regex but are not mentioned in the validation error messages

File: app/uk/gov/hmrc/agentoverseasapplicationmaintainerfrontend/views/reject_reason.scala.html [hunk 2]
```diff
@@ -50,7 +50,7 @@ <h1 class="heading-xlarge margin-bottom-10 margin-top-0">@Messages("reject_reaso
                 </div>
                 <div class="panel panel-border-narrow js-hidden" id="other-reason-panel">
                     <label class="form-label" for="other-reason">@Messages("reject_reason.reason")</label>
-                    <textarea class="form-control form-control-3-4" name="other-reason" id="other-reason" rows="5"></textarea>
+                    <textarea class="form-control form-control-3-4" name="other-reason" id="other-reason" rows="5">@rejectForm.data.getOrElse("other-reason", "")</textarea>
```
Comment (johnrevill) [inline app/uk/gov/hmrc/agentoverseasapplicationmaintainerfrontend/views/reject_reason.scala.html]: I think `@rejectForm("other-reason").value` would do the same. Same with line 66.

File: test/uk/gov/hmrc/agentoverseasapplicationmaintainerfrontend/validators/CommonValidatorsSpec.scala [hunk 3]
```diff
@@ -57,8 +57,32 @@ class CommonValidatorsSpec extends UnitSpec with EitherValues {
...
-      bind("Some other reason") shouldBe Right("Some other reason")
+      shouldAcceptFieldValue("Some other reason")
+      shouldAcceptFieldValue("Some other reason.")
+      shouldAcceptFieldValue("Some other reason 123445")
+      shouldAcceptFieldValue("Reason -")
+      shouldAcceptFieldValue("Reason ,")
+      shouldAcceptFieldValue("Reason &")
+      shouldAcceptFieldValue("Reason '")
+      shouldAcceptFieldValue("Reason @")
+      shouldAcceptFieldValue("Reason ))")
```
Comment (johnrevill) [inline test/uk/gov/hmrc/agentoverseasapplicationmaintainerfrontend/validators/CommonValidatorsSpec.scala]: (?

File: test/uk/gov/hmrc/agentoverseasapplicationmaintainerfrontend/validators/CommonValidatorsSpec.scala [hunk 4]
```diff
@@ -68,13 +92,13 @@ class CommonValidatorsSpec extends UnitSpec with EitherValues {
...
-        "testKey",
-        "error.otherReason.invalid")
+      shouldRejectFieldValueAsInvalid("**")
+      shouldRejectFieldValueAsInvalid("$$")
+      shouldRejectFieldValueAsInvalid("££")
     }
 
     "return validation error if the field exceeds 255 chars limit" in {
-      bind(randomString(256)).left.value.head.message shouldBe "error.otherReason.maxlength"
+      bind(randomString(501)).left.value.head.message shouldBe "error.otherReason.maxlength"
```
Comment (johnrevill) [inline test/uk/gov/hmrc/agentoverseasapplicationmaintainerfrontend/validators/CommonValidatorsSpec.scala]: Could have another assertion here that shows that strings of 500 characters are allowed.

File: test/uk/gov/hmrc/agentoverseasapplicationmaintainerfrontend/validators/CommonValidatorsSpec.scala [hunk 5]
```diff
@@ -68,13 +92,13 @@ class CommonValidatorsSpec extends UnitSpec with EitherValues {
...
     "return validation error if the field is invalid" in {
-      bind("**").left.value should contain only FormError(
-        "testKey",
-        "error.otherReason.invalid")
+      shouldRejectFieldValueAsInvalid("**")
+      shouldRejectFieldValueAsInvalid("$$")
+      shouldRejectFieldValueAsInvalid("££")
     }
 
     "return validation error if the field exceeds 255 chars limit" in {
```
Comment (johnrevill) [inline test/uk/gov/hmrc/agentoverseasapplicationmaintainerfrontend/validators/CommonValidatorsSpec.scala]: Test name should mention the latest limit

File: test/uk/gov/hmrc/agentoverseasapplicationmaintainerfrontend/validators/CommonValidatorsSpec.scala [hunk 6]
```diff
@@ -83,18 +107,43 @@ class CommonValidatorsSpec extends UnitSpec with EitherValues {
...
-        "testKey",
-        "error.rejectReason.invalid")
+      shouldRejectFieldValueAsInvalid("**")
+      shouldRejectFieldValueAsInvalid("$$")
+      shouldRejectFieldValueAsInvalid("££")
     }
 
     "return validation error if the field exceeds 255 chars limit" in {
-      bind(randomString(256)).left.value.head.message shouldBe "error.rejectReason.maxlength"
+      bind(randomString(501)).left.value.head.message shouldBe "error.rejectReason.maxlength"
```
Comment (johnrevill) [inline test/uk/gov/hmrc/agentoverseasapplicationmaintainerfrontend/validators/CommonValidatorsSpec.scala]: Shouldn't this reject reason still have a max length of 255?

Comment (johnrevill) [reply test/uk/gov/hmrc/agentoverseasapplicationmaintainerfrontend/validators/CommonValidatorsSpec.scala]: If so, the test name needs updating and the corresponding validation message `error.rejectReason.maxlength` message needs updating too.

File: conf/messages [hunk 7]
```diff
@@ -21,10 +21,10 @@ error.rejectReason.no-box-selected=Select if you could not confirm the money lau
 
 error.otherReason.blank=Enter a reason
 error.otherReason.maxlength=Reason must be 255 characters or less
```
Comment (johnrevill) [inline conf/messages]: Should this be 500 characters now?

File: conf/messages [hunk 8]
```diff
@@ -21,10 +21,10 @@ error.rejectReason.no-box-selected=Select if you could not confirm the money lau
...
-error.otherReason.invalid=Reason must only include letters a to z, numbers, commas, full stops, apostrophes, hyphens, ampersands and spaces
+error.otherReason.invalid=Reason must only include letters a to z, numbers, commas, full stops, apostrophes, hyphens, ampersands, round or square brackets, at (@) signs, single or double quote marks, question marks, exclamation marks, colons, semicolons and spaces
 
 error.rejectReason.maxlength=Message in the rejection notice must be 255 characters or less
```
Comment (johnrevill) [inline conf/messages]: Now 500 characters

File: test/uk/gov/hmrc/agentoverseasapplicationmaintainerfrontend/validators/CommonValidatorsSpec.scala (hunk 6 above)
Comment (johnrevill) [reply test/uk/gov/hmrc/agentoverseasapplicationmaintainerfrontend/validators/CommonValidatorsSpec.scala]: Could also have another assertion in this test showing that it happily accepts lengths up to the limit.

File: app/uk/gov/hmrc/agentoverseasapplicationmaintainerfrontend/views/reject_reason.scala.html (hunk 2 above)
Comment (rvijayan) [reply app/uk/gov/hmrc/agentoverseasapplicationmaintainerfrontend/views/reject_reason.scala.html]: Would it throw any exception if the data is unavailable?

**Pull request merged**